agent/consul/discoverychain/compile.go

```diff
@@ -382,14 +390,49 @@ func (c *compiler) newTarget(service, serviceSubset, namespace, datacenter strin
 	if service == "" {
 		panic("newTarget called with empty service which makes no sense")
 	}
-	return structs.DiscoveryTarget{
+	t := structs.DiscoveryTarget{
 		Service:       service,
 		ServiceSubset: serviceSubset,
 		Namespace:     defaultIfEmpty(namespace, c.currentNamespace),
 		Datacenter:    defaultIfEmpty(datacenter, c.currentDatacenter),
 	}
+
+	// Set default connect SNI. This will be overridden later if the service has
+	// an explicit SNI value configured in service-defaults.
+	t.SNI = c.sniForTarget(t, "")
+	return t
+}
+
+// copyAndModifyTarget will duplicate the target and selectively modify it given
+// the requested inputs.
+func (c *compiler)  copyAndModifyTarget(t structs.DiscoveryTarget, service,
+	serviceSubset, namespace, datacenter, externalSNI string) structs.DiscoveryTarget {
+	t2 := t // copy
+	if service != "" && service != t2.Service {
+		t2.Service = service
+		// Reset the chosen subset if we reference a service other than our own.
+		t2.ServiceSubset = ""
+	}
+	if serviceSubset != "" && serviceSubset != t2.ServiceSubset {
+		t2.ServiceSubset = serviceSubset
+	}
+	if namespace != "" && namespace != t2.Namespace {
+		t2.Namespace = namespace
+	}
+	if datacenter != "" && datacenter != t2.Datacenter {
+		t2.Datacenter = datacenter
+	}
+	// Reset SNI to correct value
+	t2.SNI = c.sniForTarget(t2, externalSNI)
+	return t2
+}
+
+func (c *compiler)  sniForTarget(t structs.DiscoveryTarget, externalSNI string) string {
+	return connect.ServiceSNI(t.Service, t.ServiceSubset, t.Namespace,
```

Comment:
Should this function return `externalSNI` if it's not `""`?

Reply:
Yes, I guess that is still a WIP part :)